### ISSUE TITLE
Handle OpenMetrics `..._created` lines

### ISSUE THREAD

**Issue by beorn7** [prometheus/prometheus#6541]
A target fully supporting OpenMetrics will expose `..._created` lines for each metric child. The way Prometheus currently ingests OpenMetrics, each of those will create an additional time series. The value is (mostly) constant and will therefore compress well. However, the number of time series will (almost) double, which is a significant bump in resource usage. This is in harsh contrast to the almost non-existent usefulness of those time series in the current Prometheus context.

Prometheus negotiates OpenMetrics by default (and that behavior cannot be disabled). Thus, with each target starting to support OpenMetrics, the number of time series will increase. This will come as a surprise to the operator of the Prometheus server if the OpenMetrics support is just a side effect of upgrading to a new version of a 3rd-party-supported target. For example, once K8s components expose OpenMetrics including the `..._created`, a routine K8s upgrade will suddenly start to expose all those `..._created` lines.

While technically, a metrics change is something operators should take into account when upgrading targets, I expect a lot of confusion and surprise and even monitoring outages if we don't handle the change in a more robust way. See https://github.com/prometheus/client_python/issues/438 for reactions when the Python client started to support OpenMetrics.

The currently recommended action is probably to add `metric_relabel_configs` to drop all metrics with a name ending on `_created`. This has a number of issues:
- It is opt-in, and it is so in a very non-obvious way. Even if we advertise this practice aggressively, I'd expect many operators to not notice.
- It will drop all metrics with a name ending on `_created`, even those that are not auto-created but regular metrics.
- It has a more or less significant performance impact.

I propose to handle the `..._created` lines in a different way.

The minimal option would be to automatically ignore those lines if all the following is true:
- It's an OpenMetrics exposition.
- The `..._created` line has a corresponding "proper" metric (e.g. `foobar_created` goes along with `foobar` or `foobar_total` or `foobar_sum`/`foobar_count` etc).

The above will avoid dropping regular metrics that happen to have a name ending on `_created`.

However, perhaps we can do even better and actually make use of the `..._creaed` lines: If the creation timestamp passes certain sanity checks (earlier than the scrape time, but not too much), we can artificially insert "zero" samples for counters, histograms, and summaries that spring into existence with a value greater than zero. This would finally provide a solution to the long-standing problem described in https://github.com/prometheus/prometheus/issues/1673 .

**Comment by brian-brazil** [comment]
> The minimal option would be to automatically ignore those lines if all the following is true:

One of our design aims with OpenMetrics is that the /metrics matches what ends up in Prometheus, and silently dropping time series on the floor is not in line with that. In addition you'd get a different set of series depending on whether OM or Prometheus is negotiated.

We have already discussed this in the OM meetings, and it was judged that for the small number of users where this will actually be a problem (and keeping in mind we just saved a pile of RAM from the block indexes) they can use metric_relabel_configs.

> It has a more or less significant performance impact.

It'd be about the same as anything else. Dropping metrics via metric_relabel_configs is cheap CPU wise, but all you're saving is the head RAM for a metric - not the scraping RAM.

> we can artificially insert "zero" samples for counters, histograms, and summaries that spring into existence with a value greater than zero.

That assumes that its within the head block, which is not a good assumption. It'd also mess up the average interval calculation. I think any solution here needs to involve PromQL (possibly involving the user using a different query), not scraping/tsdb.

**Comment by roidelapluie** [comment]
If I am right only counters will have a _created value.

What would be great is that we actually use those created lines:

T1 total 1
      created at 01:00
T1+10s total 2
       created at 04:00

There is a reset here, invisible in the Prometheus exposition format but visible in om.

**Comment by brian-brazil** [comment]
Counters, Summarys and Histograms have them. Using them would be good, spotting that particular case is not the main goal as if your binary is regularly restarting more often than Prometheus scrapes then you've chosen the wrong monitoring architecture.

A realistic worst case (one of Gitlab's which is very counter heavy) is a 50% increase in the number of metrics, most will have quite a bit less than that.

**Comment by beorn7** [comment]
> A realistic worst case (one of Gitlab's which is very counter heavy) is a 50% increase in the number of metrics, most will have quite a bit less than that.

Most Prometheus servers will actually not run so close to their resource limit that it will matter much. But even if only 1% of Prometheus servers suddenly toppled over, just because their operators have performed a seemingly innocuous upgrade to the next minor release of a monitored binary, it would be a very very bad user experience. Even more so if the operator then learns that the new `..._created` time series have zero value for their use case.

**Comment by beorn7** [comment]
> One of our design aims with OpenMetrics is that the /metrics matches what ends up in Prometheus, and silently dropping time series on the floor is not in line with that.

The `..._created` lines don't really represent time series data. They are metadata for which OpenMetrics chose the same syntax as for metrics.

> In addition you'd get a different set of series depending on whether OM or Prometheus is negotiated.

I'm not quite sure I understand. Is this an argument for or against my proposal? My proposal is very much in line with the idea of creating the same set of series, no matter what exposition format is negotiated. If Prometheus format is negotiated, the instrumentation library will not create `..._created` lines. That will only happen if OM is negotiated. In which case those lines should not trigger the creation of (pseudo) time series.

> It'd be about the same as anything else. Dropping metrics via metric_relabel_configs is cheap CPU wise, but all you're saving is the head RAM for a metric - not the scraping RAM.

In my experiments, I saw a noticeable increase in CPU usage. If we just ignored the `..._created` line as early as possible during ingestion, I'd expect to waste less resources.

But this is anyway a minor point. The other points against metric_relabel_configs are more relevant.

> and keeping in mind we just saved a pile of RAM from the block indexes

Which is not a good argument to justify pointless resource wastage. There will always be users with close to optimal resource utilization. They won't allocate more resource buffer just because Prometheus is now more resource efficient than previously.

> That assumes that its within the head block, which is not a good assumption. 

We have a one-hour grace period. The sanity interval I was thinking of would be more like three or four times the scrape interval, which is way shorter than the grace period and plays nicely with the average interval calculation.

**Comment by brian-brazil** [comment]
> The ..._created lines don't really represent time series data. They are metadata for which OpenMetrics chose the same syntax as for metrics.

This is not true. Metadata in OpenMetrics means something tied to an entire metric (e.g. HELP), not to individual children within it. _created is a normal sample, just like process_start_time_seconds and node_boot_time_seconds which have the same core semantics.

> Is this an argument for or against my proposal? My proposal is very much in line with the idea of creating the same set of series, no matter what exposition format is negotiated. If Prometheus format is negotiated, the instrumentation library will not create ..._created lines. 

Against. A client library supporting OpenMetrics should be exposing _created no matter what format is negotiated, given the goal to expose the same set of series for both formats.

> But this is anyway a minor point.

Yeah, CPU isn't the resource users tend to complain about.

> Which is not a good argument to justify pointless resource wastage. There will always be users with close to optimal resource utilization. They won't allocate more resource buffer just because Prometheus is now more resource efficient than previously.

That's kinda like saying that any change in resource usage is breaking. We don't consider performance to be within the stability guarantees, and it's not even a change in Prometheus in this case.

> We have a one-hour grace period. The sanity interval I was thinking of would be more like three or four times the scrape interval, which is way shorter than the grace period and plays nicely with the average interval calculation.

Having subtly different ingestion semantics depending on when exactly a scrape happens (in addition to which format is negotiated) sounds very wrong to me.

**Comment by beorn7** [comment]
> Metadata in OpenMetrics means something tied to an entire metric (e.g. HELP), not to individual children within it.

Maybe OM defines it that way. Then OM needs an additional definition for "metadata bound to an individual child". Once all the metrics are gathered by a monitoring server like Prometheus, it becomes "per timeseries" anyway as different TYPE and HELP data will inevitably show up for the same metric name.

In the following, I will actually use the word "metric" for a "child". I'll use the word "metric family" what you call "entire metric".

> _created is a normal sample, just like process_start_time_seconds and node_boot_time_seconds which have the same core semantics.

The difference is that a `..._created` line is always connected to an existing metric, while something like `process_start_time_seconds` is on its own. It therefore cannot be metadata, as there is no data it can be meta to.

I'd compare `..._created` to exemplars or to the fixed timestamp we already have. Those could also be represented as (pseudo) metrics instead of additional fields after the sample value. I think it makes way more sense to do what we are doing, namely representing them as additional fields after the sample they belong to. I would have expected the same for the timestamp of creation. I don't know why OM decided to create a (pseudo) metric for it, but even after they did, it's semantically still metadata, and we should treat it that way in Prometheus.

> A client library supporting OpenMetrics should be exposing _created no matter what format is negotiated, given the goal to expose the same set of series for both formats.

I disagree, mostly because the timestamp of creation is not a time series but metadata.

> That's kinda like saying that any change in resource usage is breaking. 

That's not what I said. I said that we shouldn't let Prometheus use more resources in a surprising way if there is no benefit for the users.

> Having subtly different ingestion semantics depending on when exactly a scrape happens (in addition to which format is negotiated) sounds very wrong to me.

We are already doing this. If a sample has a set timestamp, and that timestamp is more than an hour in the past, we don't ingest it.

**Comment by brian-brazil** [comment]
> while something like process_start_time_seconds is on its own. It therefore cannot be metadata, as there is no data it can be meta to.

It's meta to the target as a whole, though the link is not formal.

>  don't know why OM decided to create a (pseudo) metric for it, but even after they did, it's semantically still metadata, and we should treat it that way in Prometheus.

We have already had many long discussions about how to best represent all of these, and it's not something that's going to change at this point.

> I disagree, mostly because the timestamp of creation is not a time series but metadata.

It is a time series, it's a gauge. In exactly the same way that process_start_time_seconds is.

> If a sample has a set timestamp, and that timestamp is more than an hour in the past, we don't ingest it.

There's a big difference between wholesale not ingesting old data, and ingesting apparently similar data in different ways depending on whether the value of a time series indicates it's from more than a few minutes ago. That'd be at best very confusing.

**Comment by beorn7** [comment]
I disagree wholeheartedly, but trying to explain my point in even more detail is entering the area of diminishing returns, which my time budget doesn't permit.

I guess we have described the nature of our disagreement well enough at this point. I leave it to others to chime in here and break the stalemate one way or the other.

On a more general note, whenever it comes to OM, Brian and myself seem to have very fundamental disagreements. By now, I have experienced several moments where things appeared so overly obvious to me that I was utterly flabbergasted that we even had to argue about them. Given our very similar experience with Prometheus (and the concepts that lead up to it), it's deeply worrying that we come to conclusions so stunningly different.

For me, the easiest way of defuse my worries is to assume I'm missing some key piece of insight and I'm simply wrong in a way that I cannot comprehend at the moment. I'm generally willing to take that leap of faith, especially if there are multiple experienced and smart people all telling me the same thing. With the people working on OM, that seems to be the case. To confirm, I'd still like to file an issue for OM to reconsider the decision about the representation of creation timestamps. If that still results in a consensus to represent the creation timestamp metric-like, so be it. (In which case I'd still appreciate feedback from Prometheus people if we should really treat it as a time series within Prometheus.)

**Comment by roidelapluie** [comment]
My initial understanding was @beorn7 's understanding after hearing a couple of times about openmetrics. To me, _created was just an extra information that would be stored in some ways where it could be used to do fancy things on counter functions.

I was expecting prometheus to not store them as metrics, and maybe drop them.

Another thing I did not expect is that OM-Compatible exporters should also add _created as a time series to counters etc. That is unexpected.

I am not taking a strong position here, because I don't have access to the OM discussions that led to this, but out of the blue I would more likely be on @beorn7 side here.

**Comment by pieterdejaeghere** [comment]
Just a FYI as I likely bumped into this issue as predicted by @beorn7 : After a python library upgrade I noticed my Prometheus went from ~5 million metrics to ~9 million and memory requirements went from 50GB RAM to 100GB (64GB initially provisioned, which ended with a disk swap event and ultimately process killed because of out-of-memory) . Luckily it was on virtual infra, and I could upgrade immediately. Still proved an interesting sunday afternoon experience.

**Comment by juliusv** [comment]
Not having followed OM closely, the `_created` cost issue wasn't really on my radar, and it always smells to second-guess something that was decided over long discussion by a group of experts from different groups after the fact. But I can't deny that it strikes me as a very expensive deal for the Prometheus world, with no benefits for Prometheus users. Even if it is too late to change in OpenMetrics now (is it really?), it would be interesting to understand why a time series representation (rather than the kind of representation chosen for histogram exemplars) was necessary in this case. I would also be in favor of some kind of automatic mitigation of this issue on the Prometheus side. With all the work that Prometheus developers are putting into making the ingestion and storage layer more efficient, it would be a pity to wipe a lot of those efforts out again by adding a lot of new unused metrics.

**Comment by pieterdejaeghere** [comment]
![memory consumption](https://user-images.githubusercontent.com/24585008/107191548-9a7cca80-69ec-11eb-965b-fcb65a945f5e.png)
![scrape samples post relabeling](https://user-images.githubusercontent.com/24585008/107191550-9b156100-69ec-11eb-9716-2d6e40537afc.png)
![scrape_samples_scraped](https://user-images.githubusercontent.com/24585008/107191551-9badf780-69ec-11eb-81eb-63508d9855b0.png)

I've included a memory used graph (of which prometheus process utilises 78GB currently), scrape_samples_scraped and scrape_samples_post_metrics_relabling. The change of dropping ".*_created" metrics seems to drop metrics stored to the point of how it was before, but there was little drop in memory utilisation to show for it. Might still need a prometheus process restart to free up that memory (judging from the impact of a previous restart on 4 feb (104 ->80GB)). We store data (1.8TB total) for 6 months (we should probably move to a Cortex/Thanos tiered setup at some point). As we also updated our prometheus on the same change (2.21 -> 2.23) and had a label change on most metrics, I can't be 100% sure this memory utilisation is caused by the "_created" metrics at this moment. They still seem like the most likely candidate though.

**Comment by roidelapluie** [comment]
Hello, the memory usage is probably partly because of the _created lines that you have accumulated over time.

**Comment by m1keil** [comment]
My biggest issue with these is not even the performance side but the simple fact that it pollutes the metric namespace and makes it harder for folks in our team to find the metric they actually need. This also impacts front-end tools like Grafana and its metric browser since suddenly it can bring you over the 10k limitation easily.

I do realize that this is not a Prometheus issue per se but OM and Prometheus clients alignment but I'm still surprised that this is a default behavior without an opt out option on the clients with only possible way to drop this stuff at scrape time.

Adding insult to injury, nobody can actually explain what fancy things this _created metric actually provide for the average user.

**Comment by juliusv** [comment]
Agreed. I don't think we should let this leak into the Prometheus world, where it will annoy and confuse everyone, while roughly nobody will actually use it or know what it's even there for.

**Comment by roidelapluie** [comment]
I also think we should drop them, with a config option to optionally include them for power users.

**Comment by roidelapluie** [comment]
Added to dev sumit backlog.

**Comment by bwplotka** [comment]
Not sure if we should continue discussion on this thread or separate but FYI we agreed on [the recent Prometheus DevSummit](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit#bookmark=id.zcja7rewro81) to store created timestamp in some way (NOT a series). 

This means that we definitely want to skip those `_created` series (along what @beorn7 mentioned):

> The minimal option would be to automatically ignore those lines if all the following is true:
>
> 1. It's an OpenMetrics exposition.
> 2. The ..._created line has a corresponding "proper" metric (e.g. foobar_created goes along with foobar or foobar_total or foobar_sum/foobar_count etc).

The only question is what to do with created timestamp information. Where to attach it/expose it. For context, in the [proto format](https://github.com/OpenObservability/OpenMetrics/blob/main/proto/openmetrics_data_model.proto#L119) it is stored next to each counter value.

I feel I like the option @beorn7 mentioned [in the very beginning](https://github.com/prometheus/prometheus/issues/6541#issue-544489078) despite some negatives:

> However, perhaps we can do even better and actually make use of the ..._created lines: If the creation timestamp passes certain sanity checks (earlier than the scrape time, but not too much), we can artificially insert "zero" samples for counters, histograms, and summaries that spring into existence with a value greater than zero. This would finally provide a solution to the long-standing problem described in https://github.com/prometheus/prometheus/issues/1673 

We could store is as a normal sample. However, along with the "average interval" issue, inability to append too old timestamps, there is also question IF this solution is enough for Otel data model (and OM) where such timestamps is on every counter sample (e.g. imagine Prom -> Otel conversion or OTLP metrics native export as an option someday). Maybe it's good enough and one can add additional indexing/cache to finding the latest _created timestamp?

The alternatives:

* Modify TSDB like we did for exemplars, so created timestamps are a separate entities? Sounds like a lot of complexity.
* Use Metadata as mentioned [here]( https://github.com/prometheus/prometheus/issues/10164) - good enough for "latest", but then metadata is per family not series.
* Counter as a complex value - tuple of value and created timestamps - like native histograms? It feels like waste of compute / space, but could be well compressed.

**Comment by beorn7** [comment]
If we used the `_created` timestamp to insert a synthetic zero sample, it would help Prometheus use cases, but it would make it hard indeed to serve the timestamp further down (via federation, remote write, whatever will come later), but maybe that's not a big deal for now. (Plus, we could do _both_, i.e. generating the synthetic sample _and_ store the timestamp (possibly ephemerally) to add it to federation, remote write, …) Pretty much as @bwplotka has already said above.

Even in the data model where a creation timestamp is served with every counter sample, it's still IMHO metadata as it usually changes only rarely, most typically upon restart of the target (which could even change already known metadata like type and help).

From that perspective, I wouldn't put the creation timestamp into a "structured" sample value as we do it for histogram buckets and other "fields". (I do want to make all sample values "structured" as a vision for the far future. Or arguably they are already because every sample has at least a float value and a timestamp, just that we don't see it that way… but that's a story for another day.)

Which leaves us with the metadata approach. Fundamentally, I think we need to first come to terms what we consider metadata in Prometheus and how we want to handle it. There is even a [design doc](https://docs.google.com/document/d/1XiZePSjwU4X5iaIgCIvLzljJzl8lRAdvputuborUcaQ/edit) about it, but it is, as of now, an empty sheet of paper… Our traditional view was that metadata is the type and help of a metric family. Which is only true for a single exposition. Once that exposition is scraped, the scope is now also per target (because other targets can have the same metric family with different help/type) _and_ for a particular point in time because type/help can change over time. We have already failed to properly handle these additional dimensions. One solution is to fully embrace that metadata is per metric (not metric family) and that it (occasionally) might change over time for a given series (maybe establishing a new series identity, if we want to see it that way). Once we do that, we have opened up the door for other metadata that might make sense. I often think about the configured scrape interval or the actual start and end time of a series. And here we are getting very close to the `_created` timestamp from OM.

_However_ I do believe that this "proper" metadata treatment will take a while to materialize in Prometheus, so a short term solution needs to be more makeshift. It would just be cool if it didn't add more wrinkles to Prometheus as the two metadata APIs we have already added.

**Comment by Adi9235** [comment]
Hey, I am new to Prometheus so I would like to have some guidance on how to proceed for solving this issue.

**Comment by cubxxw** [comment]
hi, I look forward to participating in 2023 GSoC and I look forward to getting guidance under this issue

**Comment by ArthurSens** [comment]
@beorn7 @bwplotka the [OpenMetrics specification](https://github.com/OpenObservability/OpenMetrics/blob/v1.0.0/specification/OpenMetrics.md#counter) says:

> A MetricPoint in a Metric with the type Counter SHOULD have a Timestamp value called Created. This can help ingestors discern between new metrics and long-running ones it did not see before.

From that statement I can understand that the `_created` timestamp use-case is only for Prometheus own use, i.e. end users of Prometheus won't ever need to query this timestamp with PromQL, right? If that is correct, I don't see a point to store `_created` as a time-series, right?

I'm not entirely sure, but I think that exemplars were implemented by adding a new storage layer to Prometheus. Would it be reasonable to do the same to support `_created`?

**Comment by ArthurSens** [comment]
There is also this [document](https://docs.google.com/document/d/1P4SVXBYsDSeUThDJrI_5_aWv6mWemx4x0e0swq7vq6Q/edit#heading=h.5prvoamow70t) that proposes a change to OpenMetrics specification, changing `_created` to become metadata.

Even if the proposal gets rejected and OM stays the same, would it make sense to store _created as a new field in [metadata](https://github.com/prometheus/prometheus/blob/5d90dc0627be9ce794910ecb5f3ce55e3ac1a151/scrape/target.go#L87-L92)?

**Comment by beorn7** [comment]
The problem with the current metadata handling is that it is (a) ephemeral (just buffering the last known state) and (b) per metric family. It's perfectly possible that a counter gets recreated during the lifetime of a series so that a series could have different creation timestamps over time. Also, each counter can have a different creation timestamp, so we need to store the creation timestamp per metric (not per metric family, i.e. all metrics of the same name).

**Comment by ArthurSens** [comment]
> The problem with the current metadata handling is that it is (a) ephemeral (just buffering the last known state) and (b) per metric family. It's perfectly possible that a counter gets recreated during the lifetime of a series so that a series could have different creation timestamps over time. Also, each counter can have a different creation timestamp, so we need to store the creation timestamp per metric (not per metric family, i.e. all metrics of the same name).

Ah, alright it makes sense to me now :)

> If we used the _created timestamp to insert a synthetic zero sample, it would help Prometheus use cases, but it would make it hard indeed to serve the timestamp further down (via federation, remote write, whatever will come later), but maybe that's not a big deal for now. (Plus, we could do both, i.e. generating the synthetic sample and store the timestamp (possibly ephemerally) to add it to federation, remote write, …) Pretty much as @bwplotka has already said above.

Asking just to make sure I'm following your line of thought 😅. With the synthetic zero approach, do you have something like this in mind?

![image](https://user-images.githubusercontent.com/24193764/226924388-3837d123-3ba0-4ae8-a7b9-f00657e6ab38.png)

Regarding serving the proper timestamp via federation/remote-write... It feels strange to support something without a use case in mind 🤔. Unless I'm lacking context, I understand the timestamp's only purpose is to help functions that are prone to errors due to service restarts and that don't change with metrics being sent over federation/remote-write. I guess the safest approach is to start with synthetic zeros and, if a use-case shows up, we add support for storing the timestamp as well

**Comment by beorn7** [comment]
I'm still not convinced that storing the created timestamp in the current metadata store makes any sense. What is the use case, indeed? Maybe propagating the created timestamp via remote write is the best argument for now, but again, I think metadata in remote write is kind of wonky anyway, and maybe we shouldn't shovel even more dung on that wagon (and instead fix metadata handling fundamentally first).

Inserting a synthetic sample (in @ArthurSens's example (is it really hand-written? Beautiful…), that would indeed be a sample of value 0 at t=70s) still makes a lot of sense to me, as it is immediately useful. It doesn't "break" the scrape interval, as Prometheus deliberately doesn't depend on a fixed scrape interval (just compression works better if the scrapes are regular). I don't see a problem with out-of-order (but we can discuss implementation details later to see what devil is in the detail – I would just first like to find out if the general approach has any merit).

The synthetic sample would get propagating via remote-write, and maybe that covers the one use case, and remote-write won't even be interested in the created timestamp anymore in that case.

There is one more use case, though, that came to my mind: In principle, a counter reset can go undetected, if the counter has already gone up enough again to not go down. Example: Counter is scraped at value 10. It then gets reset. But before it gets scraped again, it is incremented all the way to 15. Prometheus will then just see an increment of 5 between the two scrapes. However, there would be a change of the created timestamp in between the two scrapes. It doesn't even have to be perfectly aligned, just be noticing the change, it would enable Prometheus to see the "unseeable" counter reset and correctly calculate an increment of 15. I do think, though, that this use case is fairly niche. It would also require Prometheus to store the counter resets separately. This is, however, one of the uses of the "metadata-over-time" store I have in mind. If we store counter resets separately, we can save a lot of computation effort when rate'ing over long ranges. This is already done for native histograms as a precedent.

**Comment by ArthurSens** [comment]
Hi everybody, we have a design document ready for review on this issue :) 

https://docs.google.com/document/d/1kakDVn8aP1JerimLeazuMfy2jaB14W2kZDHMRokPcv4/edit#heading=h.cmqufn5ktjuv

**Comment by bwplotka** [comment]
So in https://github.com/prometheus/proposals/blob/main/proposals/2023-06-13_created-timestamp.md#openmetrics-text we agreed to convert this metric to "created timestamp" data. For now with https://github.com/prometheus/prometheus/pull/12733 it will mean either new "zero sample" or in future--a metadata entry. All behind feature flags.

In the same proposal we also mentioned what happens without that feature flag, by default:

> The removal of a special metric is a breaking change. However, we still propose doing it to stop the spread of this confusing pattern in the community (with adequate warning). The current usage of created metrics is low and users are [actually not adopting OM](https://github.com/prometheus/client_python/tree/v0.17.1#disabling-_created-metrics) for that reason. Thus allowing users to opt out (users can set if they want to preserve those metrics) seems like a good way forward.

We decided to explicitly DROP such series, with extra option to bring it back as normal series if anyone depends on it. We will make sure it's very visible on release notes as well. We could also wait with this on 3.0. 🤔 

As per option to bring this back, I would propose either flag (`--scrape.om-text-kee-created-series`) or scrape configuration field similar to [ScrapeProtocols](https://github.com/prometheus/prometheus/blob/8ab14f2456683ca47978702b213ed6b60f6f8186/config/config.go#L576) per job (e.g. `OpenMetrics.KeepCreatedSeries`) 🤔 

WDYT?

**Comment by ArthurSens** [comment]
> We decided to explicitly DROP such series, with extra option to bring it back as normal series if anyone depends on it. We will make sure it's very visible on release notes as well. We could also wait with this on 3.0. 🤔

It sounds good to me if we decide to merge this only when releasing 3.0

> As per option to bring this back, I would propose either flag (`--scrape.om-text-kee-created-series`) or scrape configuration field similar to [ScrapeProtocols](https://github.com/prometheus/prometheus/blob/8ab14f2456683ca47978702b213ed6b60f6f8186/config/config.go#L576) per job (e.g. `OpenMetrics.KeepCreatedSeries`) 🤔
> 
> WDYT?

I think we would add too much complexity if we go this route. We will have a flag to enable created timestamp ingestion, then another one to allow OM's _created as another series. 

Correct me if I'm wrong, but OM's intention was never to ingest _created lines as a new series and I bet money that nobody is actively using those metrics in production. I would go with flipping the behavior to ingest those lines with the zero-injection/metadata approach and not have the option to ingest as time series. If REALLY needed, one can expose that timestamp through instrumentation

**Comment by bwplotka** [comment]
> It sounds good to me if we decide to merge this only when releasing 3.0

Did you want to say, you don't agree and you prefer to wait for 3.0? 🙃

The problem with that is then we have to wait for 3.0 to recommend SDKs to advertise _created by default (just consequence).

Plus sounds like you vote to NOT add optionality even to preserve series not matter what (on 3.0). I guess we could do that and risk some potential breaks for people (which we didn't hear any proof of).  I guess I could be happy with that as long as we DO DROP created series at least on `created-timestamp-zero-ingestion`feature flag etc.

>  If REALLY needed, one can expose that timestamp through instrumentation

What do you mean?

**Comment by bwplotka** [comment]
I propose we tackle this in [OM 2.0](https://github.com/prometheus/OpenMetrics/issues/276)

**Comment by bwplotka** [comment]
Related work: https://github.com/prometheus/docs/pull/2636
